cache/disk/disk.go

```diff
@@ -288,153 +242,191 @@ func (c *Cache) Put(kind cache.EntryKind, hash string, expectedSize int64, r io.
+		if err != nil {
 			return err
 		}
 		actualHash := hex.EncodeToString(hasher.Sum(nil))
 		if actualHash != hash {
 			return fmt.Errorf(
-				"hashsums don't match. Expected %s, found %s", key, actualHash)
+				"hashsums don't match. Expected %s, found %s", hash, actualHash)
 		}
 	} else {
 		if bytesCopied, err = io.Copy(f, r); err != nil {
 			return err
 		}
 	}
 
+	if isSizeMismatch(bytesCopied, size) {
+		return fmt.Errorf(
+			"sizes don't match. Expected %d, found %d", size, bytesCopied)
+	}
+
 	if err = f.Sync(); err != nil {
 		return err
 	}
 
 	if err = f.Close(); err != nil {
 		return err
 	}
+	closeFile = false
 
-	if bytesCopied != expectedSize {
-		return fmt.Errorf(
-			"sizes don't match. Expected %d, found %d", expectedSize, bytesCopied)
+	return nil
+}
+
+// This must be called when the lock is not held.
+func (c *Cache) commit(key string, tempfile string, finalPath string, size int64) (unreserve bool, removeTempfile bool, err error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if size > 0 && !c.lru.Unreserve(size) {
+		log.Printf("ERROR: failed to unreserve: %d", size)
+	}
+	unreserve = false
+
+	if !c.lru.Add(key, &lruItem{size: size}) {
+		err = fmt.Errorf("failed to add: %s, size %d", key, size)
+		log.Println("ERROR:", err.Error())
+		return unreserve, removeTempfile, err
 	}
 
-	// Rename to the final path
-	err = os.Rename(tmpFilePath, filePath)
-	if err == nil {
-		// Only commit if renaming succeeded.
-		// This flag is used by the defer() block above.
-		shouldCommit = true
+	err = os.Rename(tempfile, finalPath)
+	if err != nil {
+		log.Printf("ERROR: failed to rename \"%s\" to \"%s\": %v",
+			tempfile, finalPath, err)
+		log.Println("Removing", key)
+		c.lru.Remove(key)
+		return unreserve, removeTempfile, err
 	}
+	removeTempfile = false
 
-	return err
+	// Commit successful if we made it this far! \o/
+	return unreserve, removeTempfile, nil
 }
 
-// Return two bools, `available` is true if the item is in the local
-// cache and ready to use.
-//
-// `tryProxy` is true if the item is not in the local cache but can
-// be requested from the proxy, in which case, a placeholder entry
-// has been added to the index and the caller must either replace
-// the entry with the actual size, or remove it from the LRU.
-func (c *Cache) availableOrTryProxy(key string) (available bool, tryProxy bool) {
-	inProgress := false
-	tryProxy = false
+// Return a non-nil *os.File and non-negative size if the item is available
+// locally, and a boolean that indicates if the item is not available locally
+// but that we can try the proxy backend.
+func (c *Cache) availableOrTryProxy(key string, size int64, blobPath string) (rc io.ReadCloser, foundSize int64, tryProxy bool) {
+	var err error
+	var fileInfo os.FileInfo
 
 	c.mu.Lock()
+	defer c.mu.Unlock()
 
-	existingItem, found := c.lru.Get(key)
-	if found {
-		if !existingItem.(*lruItem).committed {
-			inProgress = true
+	val, available := c.lru.Get(key)
+	if available {
```

Comment:
If you move computing `tryProxy` above the `if available` block you can flatten it

```go
if available {
  item := val.(*lruItem)
  if isSizeMismatch(size, item.size) {
    return nil, -1, tryPoxy
  }
  
  fileInfo, err = os.Stat(blobPath)
  if err != nil {
    return nil, -1, tryProxy
  }
  
  foundSize := fileInfo.Size()
  if isSizeMismatch(size, foundSize) {
    return nil, -1, tryProxy
  }
  
  f, err := os.Open(blobPath)
  if err != nil {
    return nil, -1, tryProxy
  }
  
  return f, foundSize, false
}
```

I think this is a lot easier to read but also makes it clear that `err` is not being handled. IIUC at least if `os.Open()` errs we should surface this no?

Reply:
One issue with that setup is that calculating `tryProxy` will (usually) need to reserve space, but we do not want to reserve space if the blob available locally. I guess we could do that with a more complicated deferred function (maybe unreserve space, then unlock the mutex).